ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ProtocolSchedule.java

```diff
@@ -1,28 +1,56 @@
 /*
- * Copyright ConsenSys AG.
  *
- * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
- * the License. You may obtain a copy of the License at
+ *  * Copyright Hyperledger Besu Contributors.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ *  * the License. You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  * specific language governing permissions and limitations under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
  *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
- * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
- * specific language governing permissions and limitations under the License.
- *
- * SPDX-License-Identifier: Apache-2.0
  */
+
 package org.hyperledger.besu.ethereum.mainnet;
 
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.BlockHeaderBuilder;
 import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
 
-public interface ProtocolSchedule
-    extends HeaderBasedProtocolSchedule, PrivacySupportingProtocolSchedule {
+import java.math.BigInteger;
+import java.util.Optional;
+import java.util.function.Predicate;
 
-  ProtocolSpec getByBlockNumber(long number);
+public interface ProtocolSchedule extends PrivacySupportingProtocolSchedule {
 
-  @Override
-  default ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader) {
-    return getByBlockNumber(blockHeader.getNumber());
+  ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader);
+
+  default ProtocolSpec getForNextBlockHeader(
+      final BlockHeader parentBlockHeader, final long timestampForNextBlock) {
+    final BlockHeader nextBlockHeader =
+        BlockHeaderBuilder.fromHeader(parentBlockHeader)
+            .number(parentBlockHeader.getNumber() + 1)
+            .timestamp(timestampForNextBlock)
+            .parentHash(parentBlockHeader.getHash())
+            .blockHeaderFunctions(new MainnetBlockHeaderFunctions())
+            .buildBlockHeader();
+    return getByBlockHeader(nextBlockHeader);
   }
+
+  Optional<BigInteger> getChainId();
+
+  String listMilestones();
+
+  void putMilestone(
+      final boolean isTimestampMilestone,
```

Comment:
Would rather find a more encapsulated way to deal with this, ideally private to UnifiedProtocolSchedule (only used by getByBlockHeader)

Reply:
could move this into another interface like what was done with the mutable methods in the current implementation at least then it would only be exposed for the few places it's needed like the protocol schedule builder.